examples/ex18.hpp

```diff
@@ -0,0 +1,558 @@
+
+// Implementation of class DomainIntegrator
+DomainIntegrator::DomainIntegrator(const int dim) : flux(num_equation, dim) { }
+
+void DomainIntegrator::AssembleElementMatrix2(const FiniteElement &trial_fe,
+                                              const FiniteElement &test_fe,
+                                              ElementTransformation &Tr,
+                                              DenseMatrix &elmat)
+{
+   // Assemble the form (vec(v), grad(w))
+
+   // Trial space = vector L2 space (mesh dim)
+   // Test space  = scalar L2 space
+
+   const int dof_trial = trial_fe.GetDof();
+   const int dof_test = test_fe.GetDof();
+   const int dim = trial_fe.GetDim();
+
+   shape.SetSize(dof_trial);
+   dshapedr.SetSize(dof_test, dim);
+   dshapedx.SetSize(dof_test, dim);
+
+   elmat.SetSize(dof_test, dof_trial * dim);
+   elmat = 0.0;
+
+   const int maxorder = max(trial_fe.GetOrder(), test_fe.GetOrder());
+   const int intorder = 2 * maxorder;
+   const IntegrationRule *ir = &IntRules.Get(trial_fe.GetGeomType(), intorder);
+
+   for (int i = 0; i < ir->GetNPoints(); i++)
+   {
+      const IntegrationPoint &ip = ir->IntPoint(i);
+
+      // Calculate the shape functions
+      trial_fe.CalcShape(ip, shape);
+      shape *= ip.weight;
+
+      // Compute the physical gradients of the test functions
+      Tr.SetIntPoint(&ip);
+      test_fe.CalcDShape(ip, dshapedr);
+      Mult(dshapedr, Tr.AdjugateJacobian(), dshapedx);
+
+      for (int d = 0; d < dim; d++)
+      {
+         for (int j = 0; j < dof_test; j++)
+         {
+            for (int k = 0; k < dof_trial; k++)
+            {
+               elmat(j, k + d * dof_trial) += shape(k) * dshapedx(j, d);
+            }
+         }
+      }
+   }
+}
+
+// Implementation of class FaceIntegrator
+FaceIntegrator::FaceIntegrator(RiemannSolver &rsolver_, const int dim) :
+   rsolver(rsolver_),
+   funval1(num_equation),
+   funval2(num_equation),
+   nor(dim),
+   fluxN(num_equation) { }
+
+void FaceIntegrator::AssembleFaceVector(const FiniteElement &el1,
+                                        const FiniteElement &el2,
+                                        FaceElementTransformations &Tr,
+                                        const Vector &elfun, Vector &elvect)
+{
+   // Compute the term <F.n(u),[w]> on the interior faces.
+   const int dof1 = el1.GetDof();
+   const int dof2 = el2.GetDof();
+
+   shape1.SetSize(dof1);
+   shape2.SetSize(dof2);
+
+   elvect.SetSize((dof1 + dof2) * num_equation);
+   elvect = 0.0;
+
+   DenseMatrix elfun1_mat(elfun.GetData(), dof1, num_equation);
+   DenseMatrix elfun2_mat(elfun.GetData() + dof1 * num_equation, dof2,
+                          num_equation);
+
+   DenseMatrix elvect1_mat(elvect.GetData(), dof1, num_equation);
+   DenseMatrix elvect2_mat(elvect.GetData() + dof1 * num_equation, dof2,
+                           num_equation);
+
+   const int order = std::max(el1.GetOrder(), el2.GetOrder());
+   const int intorder = 2 * order + 2;
```

Comment:
I am not sure about this line. This is how it is done in DGTraceIntegrator

```   
   const IntegrationRule *ir = IntRule;
   if (ir == NULL)
   {
      int order;
      // Assuming order(u)==order(mesh)
      if (Trans.Elem2No >= 0)
         order = (min(Trans.Elem1->OrderW(), Trans.Elem2->OrderW()) +
                  2*max(el1.GetOrder(), el2.GetOrder()));
      else
      {
         order = Trans.Elem1->OrderW() + 2*el1.GetOrder();
      }
      if (el1.Space() == FunctionSpace::Pk)
      {
         order++;
      }
      ir = &IntRules.Get(Trans.FaceGeom, order);
   }
```
There, order is always odd whereas here its always even. The easiest way to test this would be just to see ir->GetNPoints(). It needs to be (polynomial order + 1) for 2D. If that is the case, then this is fine.

Reply:
@tzanio I don't think I can review the full thing for now.

Reply:
Thanks! I took that method of computing the quadrature. That's a good catch - that was old code that over-integrated the face terms.